src/Providers/Http/DTO/Request.php

```
@@ -0,0 +1,383 @@
+                return json_encode($this->data, JSON_THROW_ON_ERROR);
+            }
+
+            // Default to URL encoding for forms
+            return http_build_query($this->data);
+        }
+
+        return null;
+    }
+
+    /**
+     * Gets the Content-Type header value.
+     *
+     * @since n.e.x.t
+     *
+     * @return string|null The Content-Type header value or null if not set.
+     */
+    private function getContentType(): ?string
+    {
+        $values = $this->getHeader('Content-Type');
+        return $values !== null ? $values[0] : null;
+    }
+
+    /**
+     * Returns a new instance with the specified header.
+     *
+     * @since n.e.x.t
+     *
+     * @param string $name The header name.
+     * @param string|list<string> $value The header value(s).
+     * @return self A new instance with the header.
+     */
+    public function withHeader(string $name, $value): self
+    {
+        $headers = $this->headers;
+        $headers[$name] = is_array($value) ? array_values($value) : [$value];
+
+        return new self($this->method, $this->uri, $headers, $this->data);
+    }
+
+    /**
+     * Returns a new instance with the specified data.
+     *
+     * @since n.e.x.t
+     *
+     * @param string|array<string, mixed> $data The request data.
+     * @return self A new instance with the data.
+     */
+    public function withData($data): self
+    {
+        return new self($this->method, $this->uri, $this->headers, $data);
+    }
+
+    /**
+     * Normalizes headers to ensure they are all arrays.
+     *
+     * @since n.e.x.t
+     *
+     * @param array<string, string|list<string>> $headers The headers to normalize.
+     * @return array<string, list<string>> The normalized headers.
+     */
+    private function normalizeHeaders(array $headers): array
+    {
+        $normalized = [];
+        foreach ($headers as $name => $value) {
+            $normalized[$name] = is_array($value) ? array_values($value) : [$value];
```

Technically, we should probably also make sure there are no duplicates here. Ideally, if there is a `$name` "Content-Type" and another `$name` "content-type", their values should probably be merged? Or only the values from one should be used, or we should throw an exception.

We're already considering case insensitive lookup for the get methods, but we need to also consider it for setting headers.

---

I like your idea of a unified way of handling all this, so I made a `HeadersCollection` class, which handles these fun scenarios. Resolved in 928101e